examples/gno.land/p/gnoland/commondao/commondao.gno

```diff
@@ -0,0 +1,235 @@
+func (dao CommonDAO) Path() string {
+	parent := dao.Parent()
+	if parent != nil {
+		prefix := parent.Path()
+		if prefix != "" {
+			return prefix + PathSeparator + dao.slug
+		}
+	}
+	return dao.slug
+}
+
+// Parent returns the parent DAO.
+// Null can be returned when DAO has no parent assigned.
+func (dao CommonDAO) Parent() *CommonDAO {
+	return dao.parent
+}
+
+// TopParent returns the topmost parent DAO.
+// The top parent is the root of the DAO tree.
+func (dao *CommonDAO) TopParent() *CommonDAO {
+	parent := dao.Parent()
+	if parent != nil {
+		return parent.TopParent()
+	}
+	return dao
+}
+
+// Members returns the list of DAO members.
+func (dao CommonDAO) Members() *addrset.Set {
+	return dao.members
+}
+
+// ActiveProposals returns all active DAO proposals.
+func (dao CommonDAO) ActiveProposals() rotree.IReadOnlyTree {
+	return dao.active
+}
+
+// FinishedProposalsi returns all finished DAO proposals.
+func (dao CommonDAO) FinishedProposals() rotree.IReadOnlyTree {
+	return dao.finished
+}
+
+// Propose creates a new DAO proposal.
+func (dao *CommonDAO) Propose(creator std.Address, d ProposalDefinition) (*Proposal, error) {
+	id, ok := dao.genID.TryNext()
+	if !ok {
+		return nil, ErrOverflow
+	}
+
+	p, err := NewProposal(uint64(id), creator, d)
+	if err != nil {
+		return nil, err
+	}
+
+	key := makeProposalKey(p.ID())
+	dao.active.Set(key, p)
+	return p, nil
+}
+
+// GetActiveProposal returns an active proposal.
+func (dao CommonDAO) GetActiveProposal(proposalID uint64) (_ *Proposal, found bool) {
+	key := makeProposalKey(proposalID)
+	if v, ok := dao.active.Get(key); ok {
+		return v.(*Proposal), true
+	}
+	return nil, false
+}
+
+// GetFinishedProposal returns a finished proposal.
+func (dao CommonDAO) GetFinishedProposal(proposalID uint64) (_ *Proposal, found bool) {
+	key := makeProposalKey(proposalID)
+	if v, ok := dao.finished.Get(key); ok {
+		return v.(*Proposal), true
+	}
+	return nil, false
+}
+
+// Vote submits a new vote for a proposal.
+func (dao *CommonDAO) Vote(member std.Address, proposalID uint64, c VoteChoice) error {
+	// TODO: Support other voting options though the proposal definition
+	if c != ChoiceYes && c != ChoiceNo && c != ChoiceAbstain {
+		return ErrInvalidVoteChoice
+	}
+
+	if !dao.Members().Has(member) {
+		return ErrNotMember
+	}
+
+	p, found := dao.GetActiveProposal(proposalID)
+	if !found {
+		return ErrProposalNotFound
+	}
+	return p.record.AddVote(member, c)
+}
+
+func (dao *CommonDAO) Tally(p *Proposal) error {
+	// TODO: Support other voting options though the proposal definition
+	record := p.VotingRecord()
+	votesCount := record.VoteCount(ChoiceYes) + record.VoteCount(ChoiceNo) // Only consider YES/NO votes
+	membersCount := dao.Members().Size()
+	percentage := float64(votesCount) / float64(membersCount)
+	if percentage < p.Quorum() {
+		return ErrLowParticipation
+	}
+
+	if !p.Definition().Tally(record, membersCount) {
+		return ErrNoConcensus
+	}
+	return nil
+}
```

Comment:
this method should be on Proposal struct, or change the definition to `Tally(proposalID uint64) error` and do the lookup.